kernel/aster-nix/src/util/iovec.rs

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::mem;
+
+use super::read_val_from_user;
+use crate::{
+    prelude::*,
+    util::{read_bytes_from_user, write_bytes_to_user},
+};
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Pod)]
+pub struct IoVec {
+    base: Vaddr,
+    len: usize,
+}
+
+impl IoVec {
+    pub const fn base(&self) -> Vaddr {
+        self.base
+    }
+
+    pub const fn len(&self) -> usize {
+        self.len
+    }
+
+    pub fn read_from_user(&self, buffer: &mut [u8]) -> Result<()> {
+        debug_assert!(buffer.len() == self.len);
+        if buffer.len() != self.len {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "the length of buffer is not equal to the length of io vec"
+            );
+        }
+
+        read_bytes_from_user(self.base, buffer)
+    }
+
+    pub fn write_to_user(&self, buffer: &[u8]) -> Result<()> {
+        debug_assert!(buffer.len() == self.len);
+        if buffer.len() != self.len {
+            return_errno_with_message!(
+                Errno::EINVAL,
+                "the length of buffer is not equal to the length of io vec"
+            );
+        }
+
+        write_bytes_to_user(self.base, buffer)
+    }
+}
+
+/// Iterator over user provided Io Vectors
+#[derive(Debug, Clone)]
+pub struct IoVecIter {
+    start_addr: Vaddr,
+    iovec_count: usize,
+    current: usize,
+}
+
+impl IoVecIter {
+    /// Creates a new `IoVecIter`
+    pub const fn new(start_addr: Vaddr, iovec_count: usize) -> Self {
+        Self {
+            start_addr,
+            iovec_count,
+            current: 0,
+        }
+    }
+}
+
+impl Iterator for IoVecIter {
+    type Item = Result<IoVec>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.current < self.iovec_count {
+            let addr = self.start_addr + self.current * mem::size_of::<IoVec>();
+            self.current += 1;
+
+            let res = read_val_from_user::<IoVec>(addr);
```

Comment:
The current implementation of `IoVecIter` is wrong, as far as I can tell from [the Linux implementation](https://elixir.bootlin.com/linux/v6.8.7/source/lib/iov_iter.c#L1145). Instead of trying to copy them one by one when it is used, we need to copy all the `IoVec`s into kernel space at the beginning.
 - The latter approach actually causes problems. What error code should we return to the user application when `EFAULT` happens? The answer is inconsistent if the `EFAULT` happens on the first `IoVec` or in the middle.

Reply:
I have refactored the implementation to read all `IoVec`s at beginning. But I didn't find any error number other than `EFAULT`. (The only exception is that we return `EINVAL` if iovec.len < 0).

Reply:
> The latter approach actually causes problems.

Sorry, there is a typo. I want to mean the _former_ approach (which copies `IoVec`s one by one when they are used) causes problems, not the latter approach

.> But I didn't find any error number other than EFAULT.

Yes. If any `IoVec` cannot be accessed, we should directly fail with `EFAULT`.